internal/destination/destination.go

```diff
@@ -168,18 +170,34 @@ func (d *Destination) Open(ctx context.Context) error {
 }
 
 // Write writes a record into a Destination.
-func (d *Destination) Write(_ context.Context, records []sdk.Record) (int, error) {
-	if !d.nc.IsConnected() {
-		return 0, errWriteUnavailable
-	}
-
-	for i, record := range records {
-		if err := d.writer.Write(record); err != nil {
-			return i, fmt.Errorf("write: %w", err)
+func (d *Destination) Write(ctx context.Context, records []sdk.Record) (int, error) {
+	timeout, cancel := context.WithTimeout(ctx, d.config.RetryWait*time.Duration(d.config.RetryAttempts))
+	defer cancel()
+
+	attempts := 0
+	recorded := 0
+	for _, record := range records {
+		attempts++
+
+		select {
+		case <-timeout.Done():
+			return recorded, errWriteTimeout
+		default:
+			if err := d.writer.Write(record); err != nil {
```

Comment:
I'm thinking if we should really retry for all errors, or rather choose a subset that we know we can recover from?

Reply:
Retry for a subset means we must keep revising the NATS driver repeatedly. We could also define the ones we want to avoid retrying, but those could also change over time. 

And they are hard to pick, authentication errors, for example, could happen because of bit flipping in the network.

Reply:
That's what I was thinking of, yes, authentication errors and other things that are not recoverable. I'm not going to insist on adding this, but I do think it would be nice to think about what errors are recoverable.